platform: nrf5340: adding flash layout information

ioannisg opened the pull request:
We are introducing the flash layout information for the
nRF5340 target, to be used in memory partitioning.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

- Pushing this for feedback, while working on region_defs.h header.
flash_layout.h should introduce the flash partitioning for nRF530 APP MCU target.
- The default partitioning is meant to work with upstream Zephyr samples/tests.
- Care has been taken so this partitioning looks like upstream Zephyr flash partitioning, so TF-M is compatible with Zephyr. (But note that, at the end of the day, any upstream overlays, and/or partition manager (downstream) may override flash partitioning.)

- Secure Storage and Internal Trusted Firmware partitions are placed on the flash area that Zephyr defines as image scratch.

- **_update_**: pushed the region_defs.h. as well.

ioannisg committed:
platform: nrf5340: adding flash layout information

We are introducing the flash layout information for the
nRF5340 target, to be used in memory partitioning.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

oyvindronningstad commented on platform/ext/target/nordic_nrf/nrf5340/partition/flash_layout.h:
```diff
@@ -0,0 +1,242 @@
+ *    0x0007_8000 Non-secure image primary (480 KB)
+ * 0x000f_0000 Secure Storage Area (16 KB)
+ * 0x000f_4000 Internal Trusted Storage Area (8 KB)
+ * 0x000f_6000 NV counters area (4 KB)
```
Why the difference in size with vs without BL2? (512 B vs 4 KB)

ioannisg replied:
no difference - just missed to update this one.
fixed now

ioannisg committed:
platform: nrf5340: adding region_defs.h header

This commit introduces the region_defs.h header for
the nRF5340 target, to be used in firmware linking,
as well as memory partitioning.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

anangl commented on platform/ext/target/nordic_nrf/nrf5340/partition/region_defs.h:
```diff
@@ -0,0 +1,160 @@
+
+#include "flash_layout.h"
+
+#define BL2_HEAP_SIZE           (0x0001000)
+#define BL2_MSP_STACK_SIZE      (0x0001800)
+
+#define S_HEAP_SIZE             (0x0001000)
+#define S_MSP_STACK_SIZE_INIT   (0x0000400)
+#define S_MSP_STACK_SIZE        (0x0000800)
+#define S_PSP_STACK_SIZE        (0x0000800)
+
+#define NS_HEAP_SIZE            (0x0001000)
+#define NS_MSP_STACK_SIZE       (0x0000400)
+#define NS_PSP_STACK_SIZE       (0x0000C00)
```
Was this intentional to use 7 hex digits here? I think it's better to either use 8 (or 4) digits or omit the leading zeros, to avoid potential confusion.

ioannisg replied:
@anangl i think this was copied (and adapted) from other target porting; i noticed now that it is 7 digits in Musca, and in many other portings, as well.

oyvindronningstad replied:
+1 from me for changing to 8 or 4 digits.

ioannisg replied:
will change to 8, then

ioannisg replied:
changed.

anangl commented on platform/ext/target/nordic_nrf/nrf5340/partition/region_defs.h:
```diff
@@ -0,0 +1,160 @@
+/* NS partition information is used for SPU configuration */
+#define NS_PARTITION_START \
+            (NS_ROM_ALIAS(NS_IMAGE_PRIMARY_PARTITION_OFFSET))
+
```
Probably unwanted empty line.

ioannisg replied:
fixing this, thanks

anangl commented on platform/ext/target/nordic_nrf/nrf5340/partition/region_defs.h:
```diff
@@ -0,0 +1,160 @@
+/* Boot partition structure if MCUBoot is used:
+ * 0x0_0000 Bootloader header
+ * 0x0_0400 Image area
+ * 0x1_FC00 Trailer
```
I guess it should be `0x0_FC00`, as the [MCUBoot area size](https://github.com/NordicPlayground/trusted-firmware-m/pull/6/files#diff-1f387a5563b0b082bfcd63e10f69d8e6R77) is `(0x10000) /* 64 KB */`.

ioannisg replied:
Yes correct, good catch.

anangl commented on platform/ext/target/nordic_nrf/nrf5340/partition/flash_layout.h:
```diff
@@ -0,0 +1,242 @@
+#define SST_FLASH_AREA_ADDR     FLASH_SST_AREA_OFFSET
+/* Dedicated flash area for SST */
+#define SST_FLASH_AREA_SIZE     FLASH_SST_AREA_SIZE
+#define SST_SECTOR_SIZE         FLASH_AREA_IMAGE_SECTOR_SIZE
+/* Number of SST_SECTOR_SIZE per block */
+#define SST_SECTORS_PER_BLOCK   (0x8)
```
This would mean that the block size is `0x8000`, what is bigger than `FLASH_SST_AREA_SIZE` [defined](https://github.com/NordicPlayground/trusted-firmware-m/pull/6/files#diff-1f387a5563b0b082bfcd63e10f69d8e6R152) as `0x4000`.

ioannisg replied:
Good point, I think this was just copied as-is from other targets.
@anangl @nvlsianpu here's the docs:
```
- ``SST_SECTORS_PER_BLOCK`` - Defines the number of contiguous SST_SECTOR_SIZE
  to form a logical block in the filesystem.
```

I think we can we safely assume that a logical block in the file-system could be 4kB, so same as the SECTOR size. So I will change this to 0x1 - any complaints?

anangl replied:
No complaints from my side. Just a suggestion to do the same with `ITS_SECTORS_PER_BLOCK`.

ioannisg replied:
Yeap, done now, thanks @anangl

oyvindronningstad commented on platform/ext/target/nordic_nrf/nrf5340/partition/flash_layout.h:
```diff
@@ -0,0 +1,242 @@
+#define FLASH_AREA_0_SIZE          (FLASH_S_PARTITION_SIZE + \
+                                    FLASH_NS_PARTITION_SIZE)
+
+/* Not used*/
+#define FLASH_AREA_SCRATCH_ID      (FLASH_AREA_0_ID + 1)
+#define FLASH_AREA_SCRATCH_OFFSET  (FLASH_AREA_0_OFFSET + FLASH_AREA_0_SIZE)
+#define FLASH_AREA_SCRATCH_SIZE    (0)
```
Why is it included if not used? Please explain in comment :).

ioannisg replied:
hm, isn't it implicit? these definitions are under #else /* ! BL2 */
so without MCUboot, there's probably no need for this.

oyvindronningstad replied:
Sorry, it wasn't obvious, I meant all these 4 lines :D. It says "not used" so I'm wondering if they can all be deleted.

ioannisg replied:
So, I've got a careful look into this. I think @nvlsianpu could explain this better than me, but what I understand is that MCUboot is not using the scratch area at all, so the area is not used. However, the defines _are_ used in the flash map of BL2 as you can see in bl2/src/flash_map.c and sysflash.h.

I kept here the /* Not used*/ comment, since this is present as-is in all other target platform flash layouts.

oyvindronningstad replied:
Can you expand the comment a bit?
Also, is the bl2 flash_map.c included if `BL2` isn't defined?

ioannisg replied:
on second thoughts, @oyvindronningstad i will remove these entries completely, and will only bring them in if we notice their are useful.
Thanks for insisting on this one.

ioannisg replied:
Removed now :)

oyvindronningstad commented on platform/ext/target/nordic_nrf/nrf5340/partition/region_defs.h:
```diff
@@ -0,0 +1,159 @@
+#define NS_MSP_STACK_SIZE       (0x0000400)
+#define NS_PSP_STACK_SIZE       (0x0000C00)
+
+#define SPU_FLASH_REGION_SIZE   (0x00008000)
```
Is this define unique to us? If so, add a small comment.

ioannisg replied:
Good catch @oyvindronningstad , so I needed something for now, but I believe, this define should be taken from nrfx (FYI @anangl ) eventually. Will add a comment now, if this makes it more clear.

ioannisg replied:
added.

oyvindronningstad commented on platform/ext/target/nordic_nrf/nrf5340/partition/region_defs.h:
```diff
@@ -0,0 +1,159 @@
+#define PSA_INITIAL_ATTEST_TOKEN_MAX_SIZE   (0x250)
+
+/*
+ * SRAM SPU granularity is 8 KB on nRF5340. Alignment
```
This comment looks out of place. Can you move it or add more context?

ioannisg replied:
Actually this comment is wrong - good catch. We only care here to speak about the flash granularity, since we are defining flash partitions below. Will correct.

ioannisg replied:
updated now

oyvindronningstad commented on platform/ext/target/nordic_nrf/nrf5340/partition/region_defs.h:
```diff
@@ -0,0 +1,159 @@
+ * 0x0_0400 Image area
+ * 0x0_FC00 Trailer
+ */
+/* IMAGE_CODE_SIZE is the space available for the software binary image.
```
This and the below comment refer to IMAGE_CODE_SIZE which seems to be missing (or replaced by S and NS variants).

ioannisg replied:
I kept it as is for consistency with the other region_defs.h headers. I 'll ask upstream for the reason behind this.

oyvindronningstad replied:
I'd prefer it was referred to as something like `IMAGE_*_CODE_SIZE`, but you can leave it if you want to for the sake of consistency.

oyvindronningstad commented on platform/ext/target/nordic_nrf/nrf5340/partition/region_defs.h:
```diff
@@ -0,0 +1,159 @@
+#define IMAGE_NS_CODE_SIZE \
+            (FLASH_NS_PARTITION_SIZE - BL2_HEADER_SIZE - BL2_TRAILER_SIZE)
+
+/* The CMSE veneers shall be placed in an NSC region
+ * which will be placed in a secure SPU region defined
+ * in the beginning of the secure flash region.
```
Does this mean an entire SPU region is used for NSC veneers? That seems like a waste of space, is it required to be individually controllable via the SPU?

ioannisg replied:
Yes, it is like this unfortunately. But that's for now - we can think of a better solution after the first porting. (Actually, that SPU region is used for both the image header and the veneers)

oyvindronningstad replied:
ok thanks for the explanation.

oyvindronningstad commented on platform/ext/target/nordic_nrf/nrf5340/partition/region_defs.h:
```diff
@@ -0,0 +1,159 @@
+#endif /* BL2 */
+
+#ifndef LINK_TO_SECONDARY_PARTITION
+#define NS_IMAGE_PRIMARY_PARTITION_OFFSET (FLASH_AREA_0_OFFSET \
```
Why no secondary partition for NS?

ioannisg replied:
not sure, It does not seem to be required @oyvindronningstad 
I don't see this in other target headers

oyvindronningstad replied:
I see now they just use the secure plus an offset. let's leave it instead of creating our own system.

oyvindronningstad reviewed (commented):
Mostly nits and questions.

ioannisg commented:
@oyvindronningstad @anangl any more comments on this one?

anangl commented on platform/ext/target/nordic_nrf/nrf5340/partition/region_defs.h:
```diff
@@ -0,0 +1,160 @@
+#define NS_CODE_SIZE    (IMAGE_NS_CODE_SIZE)
+#define NS_CODE_LIMIT   (NS_CODE_START + NS_CODE_SIZE - 1)
+
+#define NS_DATA_START   (NS_RAM_ALIAS(TOTAL_RAM_SIZE / 2))
+#define NS_DATA_SIZE    (TOTAL_RAM_SIZE / 2)
```
```suggestion
#define NS_DATA_START   (NS_RAM_ALIAS(S_DATA_SIZE))
#define NS_DATA_SIZE    (TOTAL_RAM_SIZE - S_DATA_SIZE)
```
Wouldn't it be safer to define these two entries this way?

ioannisg replied:
@anangl i think your suggestion makes a lot of sense, I will apply.
(Didn't really thought so much about it; copied the solution used in the rest of the targets)

anangl replied:
AN524 [uses](https://github.com/NordicPlayground/trusted-firmware-m/blob/master/platform/ext/target/mps3/an524/partition/region_defs.h#L114) the same what I suggested.

ioannisg replied:
Applied now.
Actually, with a minor tweak, @anangl, I used 
`#define NS_DATA_START   (NS_RAM_ALIAS(S_DATA_LIMIT + 1))`

It is really weird that similar MPS2 targets have non-identical macro definition structures...

oyvindronningstad replied:
Unsure, but I think that turns out wrong. NS_RAM_ALIAS adds the offset, but S_DATA_LIMIT already includes it.

    #define NS_DATA_START (S_DATA_LIMIT + 1)

ioannisg replied:
Oh, that's probably correct @oyvindronningstad , maybe then I should revert this one.
It's really not ideal that there are no DATA_OFFSET macros for S and NS, as there are for CODE....

ioannisg replied:
the original suggestion from @anangl would work, though:
`#define NS_DATA_START   (NS_RAM_ALIAS(S_DATA_SIZE))`
Updating.
